README.md
# Dataset Inference: Ownership Resolution in Machine Learning

Repository for the paper [Dataset Inference: Ownership Resolution in Machine Learning](https://openreview.net/pdf?id=hvdKKV2yt7T) by [Pratyush Maini](https://pratyushmaini.github.io), [Mohammad Yaghini]() and [Nicolas Papernot](https://papernot.fr). This work was presented at [ICLR 2021](http://iclr.cc/Conferences/2021/) as a Spotlight Presentation.

## Additional Notes 
### Quick Start
1. Install additional dependency using `pip install typed-argument-parser`
2. Download `_model.zip` file and unzip
    ```
    └── _model
    │   └── CIFAR10-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── CIFAR10-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── CIFAR10-STL10-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── CIFAR10-STL10-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── STL10-Cat-Dog
    │   │   └── model_teacher_normalized
    │   └── STL10-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized
    ```
3. Download `_feature.zip` file and unzip
    ```
    └── _feature
    │   └── CIFAR10-Cat-Dog
    │   │   └── model_teacher_normalized-CIFAR10-Cat-Dog
    │   │   └── model_teacher_normalized-CIFAR10-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized-CIFAR10-STL10-Cat-Dog
    │   │   └── model_teacher_normalized-CIFAR10-STL10-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized-Kaggle-Cat-Dog
    │   │   └── model_teacher_normalized-STL10-Cat-Dog
    │   │   └── model_teacher_normalized-STL10-Kaggle-Cat-Dog
    ```
4. Download Kaggle Dog vs Cats train dataset `dogs-vs-cats.zip`, create `_dataset` folder and unzip into it
    ```
    └── _dataset
    │   └── dogs-vs-cats
    │   │   └── train
    │   │   │   └── cat
    │   │   │   └── dog
    ```

5. If you want to train model, please refer the command scripts in `lazy_cat_dog_train_01.sh`
6. Please use `--download_dataset` flag to download STL10 & CIFAR10 datasets, e.g:
    ```bash
    python ./src/train.py \
      --batch_size 512 \
      --mode teacher \
      --normalize 1 \
      --model_id teacher_normalized \
      --lr_mode 2 \
      --epochs 50 \
      --dataset STL10-Cat-Dog \
      --dropRate 0.3 \
      --pseudo_labels 0 \
      --use_data_parallel \
      --download_dataset
    ```
7. If you want to generate features, please refer the command scripts in `lazy_cat_dog_feature_01.sh`

8. Use `notebooks/CIFAR10_rand.ipynb` to generate result

## What does this repository contain?
Code for training and evaluating all the experiments that support the aforementioned paper are provided in this repository. 
The instructions for reproducing the results can be found below.

## Dependencies
The repository is written using `python 3.8`. To install dependencies run the command:

`pip install -r requirements.txt`
```bash
conda install pytorch==1.8.1 torchvision==0.9.1 torchaudio cudatoolkit==11.3 -c pytorch -c nvidia -c conda-forge 
conda clean --all
```

## Resolving Ownership
If you already have the extracted featured for the victim and potentially stolen models, you can proceed to inferring potential theft. A sample `jupyter notebook` to perform the same can be found at:
`src/notebooks/CIFAR10_rand.ipynb`   
You can download extracted features for our models from [this link](https://drive.google.com/drive/folders/1CLJ2a3H_oTX5b_4GLurVYoCpZUXQVpFr?usp=sharing). Save them in a directory names `files` in the root directory.

## Training your own models
`python train.py --batch_size 1000 --mode $MODE --normalize $NORMALIZE --model_id $MODEL_ID --lr_mode $LR_MODE --epochs $EPOCHS --dataset $DATASET --lr_max $LR_MAX --pseudo_labels $PSEUDO`
  > `batch_size` - Batch Size for Test Set -`default = 1000`  
  > `mode` - "Various attack strategies", type = str, default = 'teacher', choices = ['zero-shot', 'fine-tune', 'extract-label', 'extract-logit', 'distillation', 'teacher','independent','pre-act-18']  
  > `normalize`  - The normalization is performed within the model and not in the dataloader to ease adversarial attack implementation. Please take note.  
  > `model_id` - Used to compute location to load the model. See directory structure in code. 
  > `pseudo_labels` - Used in case of label only model extraction

## Generating Features
`python generate_features.py --batch_size 500 --mode $MODE --normalize $NORMALIZE --model_id $MODEL_ID --dataset $DATASET --feature_type $FEATURE`
  > `batch_size` - Batch Size for Test Set -`default = 500`  
  > `mode` - "Various attack strategies", type = str, default = 'teacher', choices = ['zero-shot', 'fine-tune', 'extract-label', 'extract-logit', 'distillation', 'teacher','independent','pre-act-18']  
  > `normalize`  - The normalization is performed within the model and not in the dataloader to ease adversarial attack implementation. Please take note.  
  > `model_id` - Used to compute location to load the model. See directory structure in code.   
  > `feature_type` - 'topgd', 'mingd', 'rand'. For black-box method use Random



doc/changes.md
# Change on author code
## Dataset Inference Code Checking Exp
- Change import in `models.py`
```python
from model_src.preactresnet import *
from model_src.wideresnet import *
from model_src.cnn import *
from model_src.resnet import *
from model_src.resnet_8x import *
# from preactresnet import *
# from wideresnet import *
# from cnn import *
# from resnet import *
# from resnet_8x import *
```
- Change model output path for `train.py`
- Change ti_500K_pseudo_labeled.pickle data path for `funcs.py`
- Change teacher model load path for `train.py`
- Fix bug for:
```
  File "/data/weijing/author-DI/src/attacks.py", line 156, in rand_steps
    remaining[remaining] = new_remaining
RuntimeError: unsupported operation: some elements of the input tensor and the written-to tensor refer to a single memory location. Please clone() the tensor before performing the operation.
```
change to `remaining[remaining.clone()] = new_remaining`
- Change model input and feature output paths for `generate_features.py`
- Add `utils/logger.py` script for print output
- Add setup `Ubuffered` output stream code for `generate_features.py` and `attack.py`
- Change `print` file direction for `generate_features.py` and `attack.py`

## Use unrelated datasets
- Add 'MINST' in `n_class` map
- Add 'MINST' in `net_mapper` map for `train.py` & `generate_features.py`
- Add new `--experiment` parser for quick setting
- Add `dataset.py` for alternate dataloading

src/params.py
import argparse
from distutils import util
import yaml
import sys

def parse_args():
    parser = argparse.ArgumentParser(description='Adversarial Training')
    ## Basics
    parser.add_argument("--config_file", help="Configuration file containing parameters", type=str)
    parser.add_argument("--dataset", help="MNIST/CIFAR10", type=str, default = "CIFAR10", choices = ["ImageNet","MNIST", "SVHN", "CIFAR10", "CIFAR100","AFAD"])
    parser.add_argument("--model_type", help="cnn/wrn-40-2/wrn-28-10/preactresnet/resnet34",
                            type=str, default = "wrn-28-10", choices = ["cnn","wrn-40-2","wrn-28-10","preactresnet", "resnet34"])
    parser.add_argument("--gpu_id", help="Id of GPU to be used", type=int, default = 0)
    parser.add_argument("--batch_size", help = "Batch Size for Train Set (Default = 100)", type = int, default = 100)
    parser.add_argument("--model_id", help = "For Saving", type = str, default = '0')
    parser.add_argument("--seed", help = "Seed", type = int, default = 0)
    parser.add_argument("--normalize", help = "Normalize training data inside the model", type = int, default = 1, choices = [0,1])
    #Only valid if restarts = 1: #0 -> Always start from Zero, #1-> Always start with 1, #2-> Start from 0/rand with prob = 1/2
    parser.add_argument("--device", help = "To be assigned later", type = str, default = 'cuda:0')
    parser.add_argument("--epochs", help = "Number of Epochs", type = int, default = 50)
    parser.add_argument("--dropRate", help = "DropRate for Teacher model", type = float, default = 0.0)
    parser.add_argument("--imagenet_architecture", help = "Imagenet Architecture", type = str, default = "wrn", choices = ["wrn","alexnet","inception"])
    
    #Threat models    
    parser.add_argument("--mode", help = "Various threat models", type = str, default = 'teacher', choices = ['zero-shot', 'prune', 'fine-tune', 'extract-label', 'extract-logit', 'distillation', 'teacher','independent','pre-act-18','random'])
    parser.add_argument("--pseudo_labels", help = "Use alternate dataset", type = int, default = 0, choices = [0,1])
    parser.add_argument("--reverse_train_test", help = "Use alternate dataset", type = int, default = 0, choices = [0,1])
    parser.add_argument("--data_path", help = "Use alternate dataset", type = str, default = None)
    parser.add_argument("--concat", help = "For Overlap Exps", type = int, default = 0, choices = [0,1])
    parser.add_argument("--concat_factor", help = "For Overlap Exps", type = float, default = 1.0)

    #LR
    parser.add_argument("--lr_mode", help = "Step wise or Cyclic", type = int, default = 1)
    parser.add_argument("--opt_type", help = "Optimizer", type = str, default = "SGD")
    parser.add_argument("--lr_max", help = "Max LR", type = float, default = 0.1)
    parser.add_argument("--lr_min", help = "Min LR", type = float, default = 0.)

    #Resume
    parser.add_argument("--resume", help = "For Resuming from checkpoint", type = int, default = 0)
    parser.add_argument("--resume_iter", help = "Epoch to resume from", type = int, default = -1)
    

    #Lp Norm Dependent
    parser.add_argument("--distance", help="Type of Adversarial Perturbation", type=str)#, choices = ["linf", "l1", "l2", "vanilla"])
    parser.add_argument("--randomize", help = "For the individual attacks", type = int, default = 0, choices = [0,1,2])
    parser.add_argument("--alpha_l_1", help = "Step Size for L1 attacks", type = float, default = 1.0)
    parser.add_argument("--alpha_l_2", help = "Step Size for L2 attacks", type = float, default = 0.01)
    parser.add_argument("--alpha_l_inf", help = "Step Size for Linf attacks", type = float, default = 0.001)
    parser.add_argument("--num_iter", help = "PGD iterations", type = int, default = 500)

    parser.add_argument("--epsilon_l_1", help = "Step Size for L1 attacks", type = float, default = 12)
    parser.add_argument("--epsilon_l_2", help = "Epsilon Radius for L2 attacks", type = float, default = 0.5)
    parser.add_argument("--epsilon_l_inf", help = "Epsilon Radius for Linf attacks", type = float, default = 8/255.)
    parser.add_argument("--restarts", help = "Random Restarts", type = int, default = 1)
    parser.add_argument("--smallest_adv", help = "Early Stop on finding adv", type = int, default = 1)
    parser.add_argument("--gap", help = "For L1 attack", type = float, default = 0.001)
    parser.add_argument("--k", help = "For L1 attack", type = int, default = 100)
    
    
    #TEST
    parser.add_argument("--path", help = "Path for test model load", type = str, default = None)
    parser.add_argument("--feature_type", help = "Feature type for generation", type = str, default = 'mingd', choices = ['pgd','topgd', 'mingd', 'rand'])

    parser.add_argument("--regressor_embed", help = "Victim Embeddings for training regressor", type = int, default = 0, choices = [0,1])

    # ! Add for quick experiment setting
    parser.add_argument("--experiment", type=str, default = "", choices = ["", "unrelated-dataset"])
    parser.add_argument("--model_dataset", help="MNIST/CIFAR10", type=str, default = "CIFAR10", choices = ["ImageNet","MNIST", "SVHN", "CIFAR10", "CIFAR100","AFAD"])

    return parser

def add_config(args):
    data = yaml.load(open(args.config_file,'r'))
    args_dict = args.__dict__
    for key, value in data.items():
        if('--'+key in sys.argv and args_dict[key] != None): ## Giving higher priority to arguments passed in cli
            continue
        if isinstance(value, list):
            args_dict[key] = []
            args_dict[key].extend(value)
        else:
            args_dict[key] = value
    return args